pkg/controller/networkpolicy/antreanetworkpolicy.go

```diff
@@ -0,0 +1,167 @@
+		if !ok {
+			klog.Errorf("Error decoding object tombstone when deleting AntreaNetworkPolicy, invalid type: %v", tombstone.Obj)
+			return
+		}
+	}
+	defer n.heartbeat("deleteANP")
+	klog.Infof("Processing AntreaNetworkPolicy %s/%s DELETE event", np.Namespace, np.Name)
+	key, _ := keyFunc(np)
+	oldInternalNPObj, _, _ := n.internalNetworkPolicyStore.Get(key)
+	oldInternalNP := oldInternalNPObj.(*antreatypes.NetworkPolicy)
+	klog.V(4).Infof("Old internal NetworkPolicy %#v", oldInternalNP)
+	err := n.internalNetworkPolicyStore.Delete(key)
+	if err != nil {
+		klog.Errorf("Error deleting internal NetworkPolicy during AntreaNetworkPolicy %s delete: %v", np.Name, err)
+		return
+	}
+	for _, atg := range oldInternalNP.AppliedToGroups {
+		n.deleteDereferencedAppliedToGroup(atg)
+	}
+	n.deleteDereferencedAddressGroups(oldInternalNP)
+}
+
+// processAntreaNetworkPolicy creates an internal NetworkPolicy instance
+// corresponding to the secv1alpha1.NetworkPolicy object. This method
+// does not commit the internal NetworkPolicy in store, instead returns an
+// instance to the caller wherein, it will be either stored as a new Object
+// in case of ADD event or modified and store the updated instance, in case
+// of an UPDATE event.
+func (n *NetworkPolicyController) processAntreaNetworkPolicy(np *secv1alpha1.NetworkPolicy) *antreatypes.NetworkPolicy {
+	appliedToGroupNames := make([]string, 0, len(np.Spec.AppliedTo))
+	// Create AppliedToGroup for each AppliedTo present in
+	// AntreaNetworkPolicy spec.
+	for _, at := range np.Spec.AppliedTo {
+		groupName := n.createAppliedToGroup(np.Namespace, at.PodSelector, at.NamespaceSelector, at.ExternalEntitySelector)
+		appliedToGroupNames = append(appliedToGroupNames, groupName)
+	}
+	rules := make([]networking.NetworkPolicyRule, 0, len(np.Spec.Ingress)+len(np.Spec.Egress))
+	// Compute NetworkPolicyRule for Egress Rule.
+	for idx, ingressRule := range np.Spec.Ingress {
+		// Set default action to ALLOW to allow traffic.
+		services, namedPortExists := toAntreaServicesForCRD(ingressRule.Ports)
+		rules = append(rules, networking.NetworkPolicyRule{
+			Direction: networking.DirectionIn,
+			From:      *n.toAntreaPeerForCRD(ingressRule.From, np, networking.DirectionIn, namedPortExists),
+			Services:  services,
+			Action:    ingressRule.Action,
+			Priority:  int32(idx),
+		})
+	}
+	// Compute NetworkPolicyRule for Egress Rule.
+	for idx, egressRule := range np.Spec.Egress {
+		// Set default action to ALLOW to allow traffic.
+		services, namedPortExists := toAntreaServicesForCRD(egressRule.Ports)
+		rules = append(rules, networking.NetworkPolicyRule{
+			Direction: networking.DirectionOut,
+			To:        *n.toAntreaPeerForCRD(egressRule.To, np, networking.DirectionOut, namedPortExists),
+			Services:  services,
+			Action:    egressRule.Action,
+			Priority:  int32(idx),
+		})
+	}
+	internalNetworkPolicy := &antreatypes.NetworkPolicy{
+		Name:            np.Name,
+		Namespace:       np.Namespace,
```

Comment:
Just thought of this when I was trying to support metrics for all 3 kinds of NetworkPolicies: How did we differentiate K8s NetworkPolicy and Antrea NetworkPolicy that have same names? I know their UIDs are different but in most places we are assuming namespace+name can identify an object at any point of time like K8s does.

Reply:
Antrea NetworkPolicy will have policyPriority while for K8s NetworkPolicy that field will be nil

Reply:
The main problem is when they are converted to internal NetworkPolicies they will override each other and generate update event instead of a new ADD event.

Reply:
given k8s NP is subset of ANP and if we cannot resolve this naming conflict easily via implementation, perhaps we could have a feature flag that let controller listens to KNP or ANP, but not both?

Reply:
the internal network policies names are UIDs no?

Reply:
I believe at the moment on agent, NP is keyed on namespace/name.

Reply:
just checked the names and UIDs are same for groups and not for the internal np.. i think we need to change the keyFunc to use UIDs..

Reply:
Not sure whether it would affect apimachinery if Namespace and Name cannot uniquely identify an entity. I didn't set UID as key but copy the UID to Name for AppliedToGroup and AddressGroup to follow the K8s convention.
I'm thinking we still need a readable and reliable way to differentiate different kinds of NetworkPolicies, maybe an approach is:
1. Copy original UID to Name like AppliedToGroup, AddressGroup 
2. Change the scope from namespace to non namespace
3. Add a NetworkPolicyReference field which points to its original NetworkPolicy UID, namespace, name and kind.

Another reason is, internal NetworkPolicy is a control plane message, which is not limited in namespace scope, and it includes ClusterNetworkPolicy which shouldn't be namespace scope too.

Reply:
proposal seems sound.. however it also has a bigger impact on the agent side.. in that case we might need to make the change in a separate PR?

Reply:
Discussed offline with Quan and agreed to make the change in a separate PR

Reply:
@tnqn : I assume you mean to use UID as Name?

Another problem to consider - in cloud VM and BM (external Node) case, we might not trust a cloud VM or a BM, and so might need some way to make sure a VM or BM Node cannot access resources not relevant to them. So, in this case, should we still use Namespace for RBAC (esp. in cloud we map a VPC to a Namespace), or we introduce some other mechanisms to isolate a Node? This might apply to CRDs agent can access as well (do we need to make a set of Namespace scoped CRDs?).
Probably we need to discuss more. @susanwu88

Reply:
@jianjuns Yes.
For cloud VM and BM case, I assume a Node is associated to a Namespace and do not require configuration of other Namespaces, at least for internal NetworkPolicies, right? If this is the case, we may keep the original Namespace.
I don't know other ways to isolate a Node when RBAC is used, as the authorization review is done by K8s API, we only know the result that whether a client is permitted to access a resource and don't know its association with Nodes or Namespaces.

Reply:
I think we may have discussed this on one of the meetings, on read/watch, any node can retrieve all NP in the same namespace, this is because NP CRD are shared across multiple nodes, and is computed at run-time.

on the other hand, each node will be given a unique role with which the node may use to access CRDs specific to itself, such as VIrtualMachine, VirtualMachineStatus, etc.

Reply:
@tnqn I think you meant internal network policy and group resources will not be Namespaced? Then how we controller agent to access the resources, unless we introduce some filtering in controller itself?

Reply:
agree.. i don't think we should eliminate the Namespace for internal network policy.. setting the `name` field of networkpolicy with UID should suffice.. and still be able to use the same keyFunc .. no? .. of course need some additional reference to store the name for other purposes.. also this reminds me our AddressGroups and AppliedToGroups are not namespaced .. same issue for agents in BM/VM?

Reply:
Ok. I think let us discuss the namespace/RBAC problems separately. It is mainly that the cloud use case increases our complexity; otherwise, ExternalEntity/Node and internal policy/group should just be cluster scope.